net/ip.c

```diff
@@ -0,0 +1,448 @@
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+#include "lib/assert.h"
+#include "lib/errno.h"
+#include "net/net_internal.h"
+#include <arpa/inet.h>          /* inet_pton, htons */
+#include <netdb.h>              /* hostent */
+#include <stdlib.h>             /* atoi */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_NET
+#include "lib/trace.h"
+
+static const char *ip_family[M0_NET_IP_AF_MAX]      = { "inet",
+							"inet6",
+							"unix" };
+
+static const char *ip_protocol[M0_NET_IP_PROTO_MAX] = { "tcp",
+							"verbs",
+							"stream",
+							"dgram",
+							"o2ib",
+							"lo" };
+
+/**
+ * Bitmap of used transfer machine identifiers. 1 is for used and 0 is for free.
+ */
+static uint8_t ip_autotm[1024] = {};
+
+/**
+ * This function convert the ip to hostname/FQDN format.
+ * Here ip is string containing ip address with numbes-and-dot notation.
+ */
+static void m0_net_ip_to_hostname(char *ip, char *hostname)
+{
+	struct hostent *he;
+	struct in_addr  addr;
+	int             rc;
+
+	inet_aton(ip, &addr);
+	he = gethostbyaddr(&addr, sizeof(addr), AF_INET);
+	M0_LOG(M0_DEBUG, "ip: %s to hostname: %s\n",ip, he->h_name);
+	rc = snprintf(hostname, M0_NET_IP_STRLEN_MAX, "%s", he->h_name);
+
+	if (rc >= M0_NET_IP_STRLEN_MAX)
+		M0_LOG(M0_ERROR, "Hostname too long. Reqd: %d buf_len: %d", rc,
+		       M0_NET_IP_STRLEN_MAX);
+}
+
+/**
+ * This function decodes the inet format address.
+ *  The inet address format is of type
+ *    <family>:<type>:<ipaddr/hostname_FQDN>@<port>
+ *    for example: "inet:tcp:127.0.0.1@3000",
+ *                 "inet:stream:lanl.gov@23",
+ *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663"
+ * Return value: 0 in case of success.
+ *               < 0 in case of error.
+ */
+static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
+{
+	uint32_t     portnum;
+	int          shift = 0;
+	int          f;
+	int          s;
+	int          rc;
+	char         ip[M0_NET_IP_STRLEN_MAX] = {};
+	char         port[M0_NET_IP_PORTLEN_MAX] = {};
+	char        *at;
+	const char  *ep_name = name;
+
+	for (f = 0; f < ARRAY_SIZE(ip_family); ++f) {
+		if (ip_family[f] != NULL) {
+			shift = strlen(ip_family[f]);
+			if (strncmp(ep_name, ip_family[f], shift) == 0)
+				break;
+		}
+	}
+	if (f >= ARRAY_SIZE(ip_family) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	for (s = 0; s < ARRAY_SIZE(ip_protocol); ++s) {
+		if (ip_protocol[s] != NULL) {
+			shift = strlen(ip_protocol[s]);
+			if (strncmp(ep_name, ip_protocol[s], shift) == 0)
+				break;
+		}
+	}
+	if (s >= ARRAY_SIZE(ip_protocol) || ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL)
+		return M0_ERR(-EINVAL);
+	else {
+		at++;
+		if (at == NULL || at[0] < '0' || at[0] > '9')
+			return M0_ERR(-EINVAL);
+		strcpy(port, at);
+		portnum = atoi(port);
+		M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
```

Comment:
1. Is zero a valid port number?
2. The other parts of the function return EINVAL, but here we just panic. Shouldn't we return EINVAL here?

Reply:
1. Is zero a valid port number?
`It is not recommended to use zero as a port num value, portnum 0 is reserved as other reserved range [0-1024]`

2. The other parts of the function return EINVAL, but here we just panic. Shouldn't we return EINVAL here?
`fixed`

Reply:
> Is zero a valid port number?
>>  It is not recommended to use zero as a port num value, portnum 0 is reserved as other reserved range [0-1024]

Let me rephrase my question: should the function `m0_net_ip_inet_parse` return -EINVAL if the user provides an address with port zero? Possible answers: 
1. yes (zero port number cannot be specified in network address);
2. no (zero port number can be specified in network address);
3. we have no specific requirements for this case.

Reply:
In net layer port number validation for reserved range is not done anywhere. Zero port number can be specified in network address, but we have no specific requirements for this case.